### PR TITLE
typeahead: line wrap improvements

### LINKED ISSUE #115164: Remote terminal local echo breaks terminal wrapping
When I have local echo enabled, this happens when wrapping to scroll. Still wraps when it doesn't need to scroll.

If I disable local echo, it works as expected

Jan-26-2021 15-44-27

### PR DESCRIPTION
**Pull request by connor4312**
Some time ago we made it so that predictions would not start for a line
until the first character was successfully predicted, in order to not
accidentally predict passwords.

This was good, except when it came to wrapping. When predicting a
character, we check the cursor position determine whether we should
insert a line wrap boundary. However if no predictions are made (or if
there's any boundary in the way) the cursor would be behind the
position for prediction purposes. Namely caused wrapping issues when
pasting long commands in the terminal.

In this commit I separate the physicalCursor from the tentativeCursor,
the latter now always tracks the position of the cursor after _all_
currently-queued predictions have been validated.

This also changed how cursors are dealt with in boundaries -- I still
push the tentative prediction, but this now only moves the tentative
cursor, not the physical one.

This existed before, so I don't think this is candidate-worthy.

Fixes #115164

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/browser/terminalTypeAheadAddon.ts [hunk 1]
```diff
@@ -510,6 +517,10 @@ class LinewrapPrediction extends NewlinePrediction implements IPrediction {
 		return ' \r';
 	}
 
+	public rollback(cursor: Cursor) {
+		return this.prevPosition ? cursor.moveTo(this.prevPosition) : '';
+	}
+
```
Comment (TylerLeonhardt) [inline src/vs/workbench/contrib/terminal/browser/terminalTypeAheadAddon.ts:523]: this appears to be the same impl as `NewLinePrediction`'s that this extends. Do you need this code?

File: src/vs/workbench/contrib/terminal/browser/terminalTypeAheadAddon.ts [hunk 2]
```diff
@@ -699,7 +710,12 @@ export class PredictionTimeline {
 	 * Cursor position -- kept outside the buffer since it can be ahead if
 	 * typing swiftly.
 	 */
-	private cursor: Cursor | undefined;
+	private _physicalCursor: Cursor | undefined;
```
Comment (TylerLeonhardt) [inline src/vs/workbench/contrib/terminal/browser/terminalTypeAheadAddon.ts:713]: "physical" here meaning "actual"? In otherwords where the cursor actually is verified by the shell?

**Comment by TylerLeonhardt** [comment]
Think this will set us up nicer for handling PowerShell's syntax highlighting?

File: src/vs/workbench/contrib/terminal/browser/terminalTypeAheadAddon.ts (hunk 2 above)
Comment (connor4312) [reply src/vs/workbench/contrib/terminal/browser/terminalTypeAheadAddon.ts:713]: That's right. The position of the cursor that the user is currently looking at on their screen (or will be looking at after all pending writes are flushed)

File: src/vs/workbench/contrib/terminal/browser/terminalTypeAheadAddon.ts (hunk 1 above)
Comment (connor4312) [reply src/vs/workbench/contrib/terminal/browser/terminalTypeAheadAddon.ts:523]: yep you're right

**Comment by connor4312** [comment]
It won't really affect it aside from making predictions more accurate in general. The paste scenario was the easy repro but I think there were other unreported bugs hiding under that rock.

File: src/vs/workbench/contrib/terminal/browser/terminalTypeAheadAddon.ts (hunk 2 above)
Comment (connor4312) [reply src/vs/workbench/contrib/terminal/browser/terminalTypeAheadAddon.ts:713]: Clarified in docblocks

**Pull request merged**